### PR TITLE
[Mono.Android] Faster native members registration

### PR DESCRIPTION
**Pull request by radekdoulik**
Fixes https://github.com/xamarin/xamarin-android/issues/1898

Measured results for XA template start-up (x86 emulator):

| RegisterNativeMembers | Before | After | After with new marshal methods |
|-----------------------|-------:|------:|-------------------------------:|
| Debug                 |  169ms |  33ms |                            N/A |
| Release               |   56ms |  18ms |                            8ms |

For more comparison, RegisterNativeMembers in the 15.8/Release takes
27ms.

The `JniTypeManager:TryRegisterNativeMembers` was quite slow, it used
reflection to find the registration method at runtime.

The new implementation finds the method during build in the Linker and
emits code to prefill a dictionary and to call the registration
methods at runtime.

That removes most of the reflection code calls, most importantly the
slow ones.

Possible future improvements:

* find better way to get the index for a given type. replace
  dictionary with perfect hash function to map type names to indexes
  or add our index to the `typemap.mj` table.

* find out a way to avoid using `MethodInfo::CreateDelegate (Type)`
  I already tried that, but wasn't able to get rid of the
  `MethodAccessException`, so had to get back to `CreateDelegate` to
  avoid that. Still think it might be possible, just need more time to
  look into it. At this point we are quite fast already though and
  there are probably more important things to optimize now. (we spend
  about 1ms in these `CreateDelegate` calls)

An example of the linker updated `Prefill` method:
```
private static void Prefill ()
{
	MagicRegistrationMap.typesMap ["xatemplateaot.MainActivity"] = 0;
	MagicRegistrationMap.typesMap ["Android.Runtime.InputStreamAdapter"] = 1;
	MagicRegistrationMap.typesMap ["Android.Runtime.JavaObject"] = 2;
	MagicRegistrationMap.typesMap ["Android.Runtime.OutputStreamAdapter"] = 3;
        ...
	MagicRegistrationMap.typesMap ["Java.Interop.TypeManager+JavaTypeManager"] = 94;
}
```

An example of the linker updated `CallRegisterMethodByIndex` method (the IL -> C# disassembler is a bit misleading at the end, where we return *false* where registration method hasn't been called):
```
private static bool CallRegisterMethodByIndex (JniNativeMethodRegistrationArguments arguments, int typeIdx)
{
	object handle;
	switch (typeIdx) {
	case 0:
		handle = typeof(MainActivity.__<$>_jni_marshal_methods).TypeHandle;
		break;
	case 1:
		handle = typeof(InputStreamAdapter.__<$>_jni_marshal_methods).TypeHandle;
		break;
	case 2:
		handle = typeof(Android.Runtime.JavaObject.__<$>_jni_marshal_methods).TypeHandle;
		break;
	case 3:
		handle = typeof(OutputStreamAdapter.__<$>_jni_marshal_methods).TypeHandle;
		break;
        ...
	case 94:
		handle = typeof(Java.Interop.TypeManager.JavaTypeManager.__<$>_jni_marshal_methods).TypeHandle;
		break;
	}
	((Action<JniNativeMethodRegistrationArguments>)Type.GetTypeFromHandle ((RuntimeTypeHandle)handle).GetMethod ("__RegisterNativeMembers").CreateDelegate (typeof(Action<JniNativeMethodRegistrationArguments>))) (arguments);
	return true;
	IL_0538:
	return false;
}
```

### CODE REVIEW

**Comment by jonpryor** [comment]
Is this Ready For Use, even experimentally?

Can we update e.g. `src/Mono.Android/Test` to make use of this, and see if the changes are visible in our Jenkins Plots?

File: src/Mono.Android/Android.Runtime/AndroidRuntime.cs [hunk 1]
```diff
@@ -248,9 +248,80 @@ static Delegate CreateDynamicCallback (MethodInfo method)
...
+
+			return rv;
+		}
+
+		class MagicRegistrationMap {
+			static Dictionary<string, int> typesMap;
+
+			static void Prefill ()
+			{
+				// fill code added by the linker
```
Comment (jonpryor) [inline src/Mono.Android/Android.Runtime/AndroidRuntime.cs]: The code that the linker will insert here should be described in the commit message, along with a suitable example.

File: src/Mono.Android/Android.Runtime/AndroidRuntime.cs [hunk 2]
```diff
@@ -248,9 +248,80 @@ static Delegate CreateDynamicCallback (MethodInfo method)
...
+
+				if (typeName == null || !typesMap.TryGetValue (typeName, out idx)) 
+					return false;
+
+				return CallRegisterMethodByIndex (arguments, idx);
+			}
+
+			static bool CallRegisterMethodByIndex (JniNativeMethodRegistrationArguments arguments, int typeIdx)
+			{
+				// updated by the linker to register known types
```
Comment (jonpryor) [inline src/Mono.Android/Android.Runtime/AndroidRuntime.cs]: The code that the linker will insert here should be described in the commit message, along with a suitable example.

**Comment by radekdoulik** [comment]
> Is this Ready For Use, even experimentally?

Yes, definitely. The Before/After column values are for XA template app without usage of the new marshal methods.

The case of using the new marshal methods is not enabled yet, as the build process changes are not committed yet. There might still be some adjustments and fixes in that area.

BTW, today I got a nice idea of doing partial (profiled) generation of the marshal methods (https://github.com/xamarin/xamarin-android/issues/2021). That would allow us to have faster startup without much added size. That should work nicely even in Debug configuration. We can include types used during normal XA template and XA/XForms template startup in `Mono.Android.dll`.

I also put down an observation about the new marshaling, where the last part which avoids us getting even better times with the new marshaling are the constructors. That still uses the old path and thus pulls in System.Reflection.Emit code. I think once we fix that, we might even get somewhere around 4ms time for `RegisterNativeMembers` during the XA template startup. (https://github.com/xamarin/xamarin-android/issues/2022)

File: src/Mono.Android/Android.Runtime/AndroidRuntime.cs (hunk 1 above)
Comment (radekdoulik) [reply src/Mono.Android/Android.Runtime/AndroidRuntime.cs]: Updated the message.

File: src/Mono.Android/Android.Runtime/AndroidRuntime.cs (hunk 2 above)
Comment (radekdoulik) [reply src/Mono.Android/Android.Runtime/AndroidRuntime.cs]: Updated the message.

**Comment by jonpryor** [comment]
Given that `Prefill()` generates statically known data, does this really need to be a `Dictionary<string, int>`?

How does the current dictionary compare to a static *sorted* array, a'la:

```csharp
static string[] TypesMap = new[]{
    "xatemplateaot.MainActivity",
    // ...
    "Java.Interop.TypeManager+JavaTypeManager",
};
```

This could be searched through in O(log n) time.  This is certainly theoretically slower than a theoretical O(1) for the Dictionary, but the Dictionary still needs to be *constructed*, which isn't "free".  (Array creation isn't free either, for that matter, but it almost certainly involves allocating less memory than the dictionary.)

For example, consider this (really stupid) [`dict-vs-array-creation.cs`](https://gist.github.com/jonpryor/9bf3ee0c36cd96cb19468ae29f2cc527) program:

```bash
$ csc dict-vs-array-creation.cs
$ mono dict-vs-array-creation.exe
create array: 00:00:01.7723561
 create dict: 00:00:11.1699148
lookup array: 00:00:06.1373072
 lookup dict: 00:00:00.2376908
```

An array with 1000 strings it's allocated in ~1/6 the time the as a Dictionary with 1000 entries.

Looking up a random value via binary search on that array takes ~30x longer than the Dictionary.  (Behold the difference between O(log n) vs. O(1)!)

In the context of Xamarin.Android, which of these is "better"?  Faster allocation should permit faster startup -- presumably? -- but there is a point where the type lookups will slow things down.  Will that slowdown *also* impact startup, or will it be a net gain (faster startup at the cost of more overhead as time goes on)?

Meanwhile, if we make *one* change to the creation metric and provide a size to the dictionary:

```csharp
    var d = new Dictionary<string, int>(1000) {
```

That *hugely* improves things:

```bash
create array: 00:00:01.9271051
 create dict: 00:00:04.9936653
lookup array: 00:00:06.1666141
 lookup dict: 00:00:00.2579320
```

Dictionary construction time is *halved*, meaning the array construction os now only ~1/3, not ~1/6, of the Dictionary time.

File: src/Mono.Android/Android.Runtime/AndroidRuntime.cs [hunk 3]
```diff
@@ -248,9 +248,80 @@ static Delegate CreateDynamicCallback (MethodInfo method)
...
+			static Dictionary<string, int> typesMap;
+
+			static void Prefill ()
+			{
+				// fill code added by the linker
+			}
+
+			static MagicRegistrationMap ()
+			{
+				typesMap = new Dictionary<string, int> ();
```
Comment (jonpryor) [inline src/Mono.Android/Android.Runtime/AndroidRuntime.cs]: ...which raises an observation: This line shouldn't be here.

Instead, `Prefill()` should do the `new Dictionary<string, int>()` execution, as `Prefill()` -- or the code which generates `Prefill()` -- *knows how big the dict should be*, and thus can *instead* generate `new Dictionary<string, int>(Count)`.

**Comment by radekdoulik** [comment]
> Given that Prefill() generates statically known data, does this really need to be a Dictionary<string, int>?

It doesn't and as I wrote in the commit message, there are better ways to do it, as the perfect hashing for example. The current version with the dictionary works reasonably well, the 'Prefill' takes 1ms in the XA template I measured.

Moving the dictionary creation to `Prefill` is one of the adjustments I have in mind. I wanted to get the whole patch out as I spent a lot time on it and wanted things to move along.

I can update the dictionary initialization in the new PR. Calling the constructor with the size is nice improvement, good idea!

Meanwhile I would like to fix cecil (or linker/jnimarshalmethod-gen?) problem, where we are breaking virtual classes and getting `System.TypeLoadException` on linked assemblies.

**Pull request merged**